### DISCUSSION TITLE
2024年开源操作系统训练营项目二：ArceOS 宏内核周报

### DISCUSSION THREAD

**Discussion by Azure-stars** [Starry-OS/.github#1] [General]
本讨论区用于收集训练营阶段四项目二：ArceOS 宏内核的周报与讨论内容。

本阶段指导文档详见：https://thu-sast9.feishu.cn/docx/ICf2dVcZ5ouH86x2G2oc4GFlnCd

**Comment by Azure-stars** [comment]
请大家在这里描述自己的选题方向、工作进展和对应的疑问（若有），格式类似如下：

- 选题：starry-next 完善
- 仓库：https://github.com/Azure-stars/Starry-On-ArceOS
- 进展：成功读取 disk.img 并且运行测例，这里有附图
- 疑问：如何去实现 clone syscall，有没有什么思路

**Comment by dynamic-pigeon** [comment]
选题：posix 接口相关

仓库：https://github.com/chy669086/futex

进展：上次写的胶水层太差了，既然例子是 futex，就直接写 futex 了。实现了 private key 以及基本的 wake wait requeue 操作。

疑问：
1. futex 有一个最大等待时间，应该怎么定义一个计时器去唤醒特定任务？
2. futex 有一个要求是使用原子操作读取一个地址上的数值，但是传入的是 u32 指针，有什么方法可以用原子操作去读这个指针吗（现在是直接转成 *mut AtomicU32 了，感觉是 ub，但是不知道其他方法了）？

**Reply by Azure-stars** [reply]
1. 可以使用轮询的思想，当进入 futex 的时候记录当前的时间戳，之后每次睡眠唤醒之后查询当前的时间戳，计算两者之间的差值从而得到当前经过的时间，可以参考 [Starry 的实现](https://github.com/Starry-OS/Starry-New/blob/main/modules/axprocess/src/futex.rs#L85)
2. 原子操作主要就是要求禁止读取过程中被打断，你可以使用 kernel_guard 来调用相关的方法，示例为[this](https://github.com/Starry-OS/Starry-New/blob/main/modules/axtrap/src/trap.rs#L4)。这里调用了 guard，保证禁用了抢占，我们可以同时禁止抢占和时间中断的出现。另外引入了这个 crate 就意味着模块调用者必须实现相关的 trait，具体的实现原理可以看 [kernel_guard](https://github.com/arceos-org/kernel_guard)。
当然也有另外一个方法，就是去看 C 语言中的实现方式，并且在 rust 里面尝试复现，即使是 unsafe 的也没有问题

**Comment by BattiestStone4** [comment]
选题：starry-next 完善
仓库：https://github.com/BattiestStone4/Starry-On-ArceOS
进展：实现简单的clone和wait系统调用。实现getppid调用。
未实现的系统调用列表更新如下：
| syscall_id | syscall_name |
| ---------- | ------------ |
| 34         | mkdirat      |
| 40         | mount        |
| 56         | openat       |
| 59         | pipe2        |
| 153        | times        |
| 160        | uname        |
| 214        | brk          |
| 221        | exec         |

220 clone（目前不执行子函数功能）
![image](https://github.com/user-attachments/assets/84e1cc92-1c91-45c4-84eb-51397ffe9fd9)

260 wait4（在别的测例上仍有问题，如waitpid会error）
![image](https://github.com/user-attachments/assets/b90e302b-bcf1-46ba-9f4c-002305b404bb)

173 getppid
![image](https://github.com/user-attachments/assets/632f77c0-6b27-438a-98a9-c5873c691b3b)

疑问：暂无，等到完善wait和clone调用时再提出。

**Reply by Azure-stars** [reply]
可以参考[这里](https://github.com/orgs/Starry-OS/discussions/1#discussioncomment-11595692)的讨论，这位同学提出了之前实现的 FD_TABLE resource 存在问题，在 new thread local 之后资源并不会是独占的，而是共享的，会导致出现问题。解决方式可以是 [修改了"thread-local"全局命名空间的实现](https://github.com/Azure-stars/Starry/commit/f0f4a3c24924bea20e40d7532499f4eb793bba3e)，另外他实现的其他系统调用也可以参考～

**Comment by xingmin1** [comment]
##### 选题：starry-next 完善
###### 仓库：

- https://github.com/xingmin1/Starry-On-ArceOS/tree/oscomp
- https://github.com/xingmin1/Starry/tree/monolithic

###### 进展：

1. [clone](https://github.com/xingmin1/Starry-On-ArceOS/commit/6ed611e93a787d1b7f2d6bf23bcfec1b597e026a)
[增加 AddrSpace 克隆功能以及WeakAxTaskRef别名类型](https://github.com/Azure-stars/Starry/commit/a288d44a58fc56982716e8d23ec91c3a688b1d0e)
2. [wait4](https://github.com/xingmin1/Starry-On-ArceOS/commit/a7676a29557f795332c26ac795fbd4c6f27782c2)
[公共化 TaskState 枚举并添加任务状态和退出码的访问方法，为wait4提供支持](https://github.com/Azure-stars/Starry/commit/72979edf46a68169e4306f6271f2354fd7e7322f)
3. [execve](https://github.com/xingmin1/Starry-On-ArceOS/commit/306e941ccb6c2ef5043f9f157509918d97a5e1e0)
4. [getppid](https://github.com/xingmin1/Starry-On-ArceOS/commit/2f5d56683540935c8bd683d4c9fe56b5281720c4)
5. [getcwd](https://github.com/Azure-stars/Starry-On-ArceOS/commit/75b456cef73581e5248e0f64e5c2577fc80cacef)
6. [实现新建任务的命名空间初始化](https://github.com/xingmin1/Starry-On-ArceOS/commit/791d43c440acd2c3fd93d22204796f27319b9876)
[修改了"thread-local"全局命名空间的实现](https://github.com/Azure-stars/Starry/commit/f0f4a3c24924bea20e40d7532499f4eb793bba3e)
[添加 FD_TABLE 和 CURRENT_DIR_PATH 的copy_inner方法](https://github.com/Azure-stars/Starry/commit/bae071f03c071a79d14cc3c1ee76ada01a76c766)
7. [sys_pipe2和sys_close](https://github.com/xingmin1/Starry-On-ArceOS/commit/402ce57d98e27a9494e23b21e3c607a831e8b289)
8. [sys_dup和sys_dup3](https://github.com/xingmin1/Starry-On-ArceOS/commit/a28b8ba984f3e38a1e67fe3bbda2368156005ee2)
9. [sys_chdir和部分sys_mkdirat](https://github.com/xingmin1/Starry-On-ArceOS/commit/7b6b99ab0d12ec78cbb40dec07ecf9d1428739c8)

###### 疑问：

1. uctx好像只用了一次，作为字段是不是有些浪费？可不可以直接写在闭包里面？
2. 原仓库的ns好像，在new_thread_local()后创建的ns里面的ax_res会直接和内核的是同一个状态，共享同一个资源，所以我改了一下，不知道有没有理解、使用错误？
3. ns 在 drop 时对 base 指向的数据没有调用 drop，是不是要在某处增加一下这个操作？
4. 由于fdtable所依赖的[(https://img.shields.io/crates/v/flatten_objects)](https://crates.io/crates/flatten_objects)不支持clone，所以fdtable的 copy 实现有些呆，有没有更好的实现思路？

**Reply by Azure-stars** [reply]
1. 我觉得没有问题，目前只是用于初始化的工作。你可以直接进行修改
2. 第二个的 axns 在 new_thread_local 之后，每一个任务会有一个独立的 area 作为 namespace，用于存放 resource 信息。虽然在复制的时候是同一个状态，但是之后操作的时候是独有的，不应当出现共享？
3. base 的 namespace 中存储的内容如果使用的是 resource 的话，应当就是使用 Arc 指针存储的，当 Arc 指针被销毁了，那么对应的 res 数据应当会自动调用 drop。
4. 可以考虑将 flatten object 的实现换为其他的数据结构，比如 vec 或者其他内容
